### ISSUE TITLE
[Question] Correct way to reference a resolver from another resolver

### ISSUE THREAD

**Issue by samdenty99** [19majkel94/type-graphql#106]
I'm attempting to create a `stats` field that has asynchronous fields, but I'm struggling on how to implement it / link it together. What would be the best way to accomplish something like below?

```ts
@ObjectType()
class Stats {
  @Field()
  totalMessages: number

  @Field()
  onlineGuests: number
}

@Resolver(of => Stats)
class StatsResolver implements ResolverInterface<Stats> {
  totalMessages() {
    return 10
  }

  async onlineGuests() {
    return 2
  }
}

@Resolver()
class StatsQuery {
  @Query(returns => Stats)
  stats = StatsResolver
}
```

**Comment by 19majkel94** [comment]
Yes, your resolver class can have simple methods that act as a helpers, so you need to annotate all your queries, mutation and field resolver methods with decorators :wink:

**Comment by samdenty99** [comment]
@19majkel94 How would you go about implementing deeply nested async query resolvers? eg.

```gql
query {
  test(a: "b") {
    test2(a: "b") {
      test3(a: "b") {
        test4
        test5
      }
    }
  }
}
```

**Comment by 19majkel94** [comment]
In the same way like all other GraphQL implementation - `test` is the Query that returns an object type that has field `test2` that returns an object type with field `test3`, etc.
That's exactly what's the schema describes and it's conformant to GraphQL specification.

I think that you are trying to do some weird stuff like in #91 - don't do that, keep your schema with queries and mutations flat, use decorators and middlewares to attach some behavior like `admin` guarded mutations, etc.

**Comment by samdenty99** [comment]
@19majkel94 Thanks! I've got the schema working correctly, it's just the `FieldResolver`'s that I'm not sure to handle.

For example:

```ts
@FieldResolver()
test() {
  return {
    // This doesn't work, the function is just converted to a string
    async doSomeOtherIntensiveQuery() {
      // ...
      return await doIntensiveQuery()
    },
    
    // This does work, but it can't do anything async
    get someIntensiveSyncQuery() {
      return intensiveSyncQuery()
    }
  }
}
```

As far as I can tell, there's no way of adding optionally computed async fields without pre-computing every single sub-field when the `FieldResolver` is called. I was asking the above to see if there was a way of returning a Resolver from within a FieldResolver (as they support async Fields)

---

In my application we serve requests that may have a `messages` field. Getting the `name` & `id` fields can be done quickly, but the messages field takes time to query the database. 

We could move all the query to the top-level, but it just wouldn't make sense for our datastructure (and would lead to more querying).

```gql
query {
  server(id: "436136130035843082") {
    name
    channels {
      id
      messages {
        timestamp
      }
    }
    channel(id: "442843083391107082") {
      name
      messages {
        content
      }
    }
  }
}
```

**Comment by 19majkel94** [comment]
My question is why you need to have `channel(id: "442843083391107082")` be defined as the `server` field, not a separate query?

So this is the GraphQL schema that works with the query you've showed:
```gql
type Query {
 server(id: ID!): Server
}

type Server {
  name: String!
  channels: [Channel!]!
  channel(id: ID!): Channel
}

type Channel {
  id: ID!
  name: String!
  messages: [Message!]!
}

type Message {
  timestamp: Int!
  content: String!
}
```

So you need to create `Server`, `Channel`, `Message` object type classes with all the fields and then create resolver class for `Server` with `channels` and `channel` field resolvers.

You shouldn't use object type as an agregate on other query for attaching metadata.
I think that `channel(id: ID!): Channel` should be a top level query as it's not logically coupled with `Server`.

So the query would look like this:
```gql
query {
  server(id: "436136130035843082") {
    name
    channels {
      id
      messages {
        timestamp
      }
    }
  }
  channel(id: "442843083391107082") {
    name
    messages {
      content
    }
  }
}
```

**Comment by 19majkel94** [comment]
No, your resolvers can't create inline deeply nested object with own resolvers. You need to go with GraphQL specification semantic and create ObjectType classes with own classes with field resolvers methods.

**Comment by 19majkel94** [comment]
@samdenty99 
So you want to "return a resolver from a FieldResolver" but how this should be reflected in schema since in GraphQL `doSomeOtherIntensiveQuery` would be a field of ObjectType returned from `test` field? Basically, it should automatically generate name based on base object type class and a field, like `PersonAgeResponse`?

**Comment by cnebs** [comment]
What if I have two mongoose models: User and ProgramTemplate.

User has a field `programTemplates` with an array of references to ProgramTemplates.

ProgramTemplate has a field `owner` with a ref to a User, and has many fields with nested refs throughout.

I have a ProgramTemplate resolver with field resolvers that resolves most fields.

If I want to resolve the `programTemplates` field on User, do I have to re-write all of the field resolvers that existed on ProgramTemplate just to get its nested values?

Sometimes I want to query (from User) details from a ProgramTemplate that exists on the user's array of `programTemplates`.
```query MyProgramTemplates {
    User {
      ... on Error {
        message
      }
      ... on User {
        programTemplates {
          name
          description
          type
          cycleLength
          daysPerWeek
          lifts {
            name
          }
        }
      }
    }
  }```
  
  Sometimes I want to query (from ProgramTemplate) the details of User that exist on the template's `owner` field.
  ```query DefaultRoutineSummaries($type: String!) {
    defaultProgramTemplates(type: $type) {
      ... on Error {
        message
      }
      ... on ProgramTemplates {
        programTemplates {
          name
          owner {
            username
          }
          description
          cycleLength
          daysPerWeek
          lifts {
            name
          }
        }
      }
    }
  }```
  
  Is there a way for me to use the User resolver on the ProgramTemplate `owner` FieldResolver?
Is there a way for me to use the ProgramTemplate resolver on the User `programTemplates` FieldResolver?
